yorkie/src/main/kotlin/dev/yorkie/document/Document.kt

```diff
@@ -369,21 +370,25 @@ public class Document(public val key: Key, private val options: Options = Option
     /**
      * Triggers an event in this [Document].
      */
-    internal suspend fun publish(event: Event) {
-        when (event) {
-            is Others.Watched -> {
-                presences.first { event.changed.actorID in it.keys }
-            }
-
-            is Others.Unwatched -> {
-                presences.first { event.changed.actorID !in it.keys }
-            }
+    internal suspend fun publishPresenceEvent(event: Event.PresenceChange) {
+        val predicate: (Presences) -> Boolean = { presences ->
+            when (event) {
+                is MyPresence.Initialized -> event.initialized == presences
+                is MyPresence.PresenceChanged -> {
+                    val actorID = event.changed.actorID
+                    actorID !in presences || event.changed.presence == presences[actorID]
+                }
 
-            is MyPresence.Initialized -> {
-                presences.first { event.initialized == it }
+                is Others.Watched -> event.changed.actorID in presences
+                is Others.Unwatched -> event.changed.actorID !in presences
+                is Others.PresenceChanged -> {
+                    val actorID = event.changed.actorID
+                    actorID !in presences || event.changed.presence == presences[actorID]
+                }
             }
-
-            else -> {}
+        }
+        withTimeoutOrNull(3_000) {
```

Comment:
do we need the timeout?

Reply:
It is logically unnecessary, but I added it just in case of an unexpected issue which can lead to an indefinite delay in the `first()` function.
I'll remove it if it looks redundant.